nca/NetworkConfig/NetworkConfigQueryRunner.py

```diff
@@ -117,59 +119,115 @@ def _execute_pair_configs_query(self, query_type, config1, config2, cmd_line_fla
                 if config1 != config2:
                     query_res, query_output, query_not_executed = self._execute_pair_configs_query(
                         self.query_name, self._get_config(config1), self._get_config(config2))
-                    res += query_res
-                    output += query_output + '\n'
-                    queries_not_executed += query_not_executed
+                    res, output, queries_not_executed = \
+                        self._update_query_results_after_one_iteration(res, query_res, output, query_output,
+                                                                       queries_not_executed, query_not_executed)
+
+        return self._return_final_query_results(res, output, queries_not_executed)
+
+    # following 3 def-s are for avoiding code repetition in previous 4 def-s
+    def _init_query_results_values(self):
+        """
+        initializes the query variable results as following
+        res : 0 (the numerical result)
+        output : will be initialized to an empty list if the output format is json, otherwise, empty str
+        (the formatted explanation result of the query)
+        queries not executed : 0 (the number of not executed queries)
+        rtype: int , Union[list, str], int
+        """
+        # json results will be appended in a list, and finally be dumped into string to
+        # ensure all results are written under one top level object to get a fixed json output format
+        output = [] if self.output_configuration.outputFormat == 'json' else ''
+        return 0, output, 0
+
+    def _update_query_results_after_one_iteration(self, result, iter_res, output, iter_output, num_not_exec, iter_not_exec):
+        """
+        gets the query result from one iteration on the configs_array, and updates the general results variables
+        with them as following:
+        1- adds the iteration numerical result to the existing numerical result
+        2. append the output to the existing one as following, if output format is json, it loads the str and appends
+        the result to the existing list (to form a json output with one top level object at last),
+        otherwise, appends to existing str
+        3. adds the iter_not_exec to the existing number of not executed queries
+        :param int result: the numerical result from running previous iterations
+        :param int iter_res: the numerical result from running this iteration
+        :param Union[list, str] output: the output from previous iterations
+        :param str iter_output: the query output from last iteration
+        :param int num_not_exec: the number of not executed queries from previous iterations
+        :param int iter_not_exec: the query_not_executed result from last iteration (0/1)
+        rtype: int , Union[list, str], int
+        """
+        result += iter_res
+        if self.output_configuration.outputFormat == 'json':
+            output.append(json.loads(iter_output))
```

Comment:
so for the json case , we get the output in string(from json) format, then convert it back into a dict to be appended to to the list, and eventually use conversion back to string of the entire list of dict objects? 
Can we instead return the output in the dict object originally ? (could also be relevant for yaml) so we don't have to use` json.loads` here?

Reply:
Hi, I am making relevant fixes,
I want to change the format of yaml output to be  a "list" of query results , instead of multiple yaml objects separated with "---" 
e.g.:
instead of :
> 
```
query: equiv_all_range1_yaml
configs:
- equiv-ranges-writing-games/kube-system/testcase16-simple-all-range
- equiv-ranges-writing-games/kube-system/testcase16-nets-all-range-partition2-nets
numerical_result: 0
textual_result: equiv-ranges-writing-games/kube-system/testcase16-simple-all-range
  and equiv-ranges-writing-games/kube-system/testcase16-nets-all-range-partition2-nets
  are semantically equivalent.
---

query: equiv_all_range1_yaml
configs:
- equiv-ranges-writing-games/kube-system/testcase16-simple-all-range
- equiv-ranges-writing-games/kube-system/testcase16-nets-all-range-partition2-notNets
numerical_result: 0
textual_result: equiv-ranges-writing-games/kube-system/testcase16-simple-all-range
  and equiv-ranges-writing-games/kube-system/testcase16-nets-all-range-partition2-notNets
  are semantically equivalent.
---
```

the result will be:

```
- query: equiv_all_range1_yaml
  configs:
  - equiv-ranges-writing-games/kube-system/testcase16-simple-all-range
  - equiv-ranges-writing-games/kube-system/testcase16-nets-all-range-partition2-nets
  numerical_result: 0
  textual_result: equiv-ranges-writing-games/kube-system/testcase16-simple-all-range
    and equiv-ranges-writing-games/kube-system/testcase16-nets-all-range-partition2-nets
    are semantically equivalent.

- query: equiv_all_range1_yaml
  configs:
  - equiv-ranges-writing-games/kube-system/testcase16-simple-all-range
  - equiv-ranges-writing-games/kube-system/testcase16-nets-all-range-partition2-notNets
  numerical_result: 0
  textual_result: equiv-ranges-writing-games/kube-system/testcase16-simple-all-range
    and equiv-ranges-writing-games/kube-system/testcase16-nets-all-range-partition2-notNets
    are semantically equivalent.
```

I want to know if it is okay for you too?

this will also be consistent with ConnectivityMap and SemanticDiff yaml results - 
and then I can dump results of all queries only once at the end in the NetworkConfigQueryRunner.py

Reply:
yes, looks good to me.

Reply:
@zivnevo can you also confirm this output change is ok?